File: common/src/models/query.rs
//! Query data model representing SQL queries

use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use super::{Condition, Join};

/// Represents a SQL query with all its components
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Query {
    /// Unique identifier for the query
    pub query_id: String,
    /// Original SQL string
    pub sql: String,
    /// Type of query (SELECT, INSERT, UPDATE, DELETE)
    pub query_type: QueryType,
    /// Columns to select (empty for SELECT *)
    pub select_columns: Vec<String>,
    /// Tables to query from
    pub from_tables: Vec<String>,
    /// WHERE conditions (optional)
    pub where_conditions: Option<Vec<Condition>>,
    /// JOIN clauses (optional)
    pub joins: Option<Vec<Join>>,
    /// Additional metadata
    pub metadata: HashMap<String, String>,
}

/// Types of SQL queries supported
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum QueryType {
    Select,
    Insert,
    Update,
    Delete,
}

impl Query {
    /// Create a new query with a generated ID
    pub fn new(sql: String, query_type: QueryType) -> Self {
        Self {
            query_id: format!("query_{}", chrono::Utc::now().timestamp_millis()),
            sql,
            query_type,
            select_columns: Vec::new(),
            from_tables: Vec::new(),
            where_conditions: None,
            joins: None,
            metadata: HashMap::new(),
        }
    }

    /// Create a new query with a specific ID
    pub fn with_id(query_id: String, sql: String, query_type: QueryType) -> Self {
        Self {
            query_id,
            sql,
            query_type,
            select_columns: Vec::new(),
            from_tables: Vec::new(),
            where_conditions: None,
            joins: None,
            metadata: HashMap::new(),
        }
    }

    /// Add a WHERE condition
    pub fn add_condition(&mut self, condition: Condition) {
        self.where_conditions
            .get_or_insert_with(Vec::new)
            .push(condition);
    }

    /// Add a JOIN clause
    pub fn add_join(&mut self, join: Join) {
        self.joins.get_or_insert_with(Vec::new).push(join);
    }

    /// Check if this is a SELECT query
    pub fn is_select(&self) -> bool {
        matches!(self.query_type, QueryType::Select)
    }

    /// Check if this query has WHERE conditions
    pub fn has_conditions(&self) -> bool {
        self.where_conditions.as_ref().map_or(false, |c| !c.is_empty())
    }

    /// Check if this query has JOINs
    pub fn has_joins(&self) -> bool {
        self.joins.as_ref().map_or(false, |j| !j.is_empty())
    }
}

File: common/src/main/java/com/distributed/sql/common/models/Query.java
package com.distributed.sql.common.models;

import java.time.LocalDateTime;
import java.util.*;

/**
 * Represents a SQL query with all its components
 */
public class Query {
    private String queryId;
    private String sql;
    private QueryType queryType;
    private List<String> selectColumns;
    private List<String> fromTables;
    private List<Condition> whereConditions;
    private List<Join> joins;
    private Map<String, String> metadata;

    public Query() {
        this.selectColumns = new ArrayList<>();
        this.fromTables = new ArrayList<>();
        this.whereConditions = new ArrayList<>();
        this.joins = new ArrayList<>();
        this.metadata = new HashMap<>();
    }

    public Query(String sql, QueryType queryType) {
        this();
        this.queryId = "query_" + System.currentTimeMillis();
        this.sql = sql;
        this.queryType = queryType;
    }

    public Query(String queryId, String sql, QueryType queryType) {
        this();
        this.queryId = queryId;
        this.sql = sql;
        this.queryType = queryType;
    }

    public void addCondition(Condition condition) {
        this.whereConditions.add(condition);
    }

    public void addJoin(Join join) {
        this.joins.add(join);
    }

    public boolean isSelect() {
        return queryType == QueryType.SELECT;
    }

    public boolean hasConditions() {
        return !whereConditions.isEmpty();
    }

    public boolean hasJoins() {
        return !joins.isEmpty();
    }

    // Getters and Setters
    public String getQueryId() {
        return queryId;
    }

    public void setQueryId(String queryId) {
        this.queryId = queryId;
    }

    public String getSql() {
        return sql;
    }

    public void setSql(String sql) {
        this.sql = sql;
    }

    public QueryType getQueryType() {
        return queryType;
    }

    public void setQueryType(QueryType queryType) {
        this.queryType = queryType;
    }

    public List<String> getSelectColumns() {
        return selectColumns;
    }

    public void setSelectColumns(List<String> selectColumns) {
        this.selectColumns = selectColumns;
    }

    public List<String> getFromTables() {
        return fromTables;
    }

    public void setFromTables(List<String> fromTables) {
        this.fromTables = fromTables;
    }

    public List<Condition> getWhereConditions() {
        return whereConditions;
    }

    public void setWhereConditions(List<Condition> whereConditions) {
        this.whereConditions = whereConditions;
    }

    public List<Join> getJoins() {
        return joins;
    }

    public void setJoins(List<Join> joins) {
        this.joins = joins;
    }

    public Map<String, String> getMetadata() {
        return metadata;
    }

    public void setMetadata(Map<String, String> metadata) {
        this.metadata = metadata;
    }
}
